logfire/_internal/instrument.py

```diff
@@ -20,118 +22,117 @@
+)
 
-    module = inspect.getmodule(func_code)
-    assert module is not None
-    filename = inspect.getsourcefile(func_code)
 
-    # We have to process the entire source file, not just the function definition,
-    # so that the compiled code has the correct context for things like closures.
-    file_source_lines, _ = inspect.findsource(func_code)
-    assert filename is not None
-    file_source = ''.join(file_source_lines)
-    tree = ast.parse(file_source)
-    transformer = InstrumentTransformer(args, logfire_name, filename, module.__name__, func_code.co_firstlineno)
-    tree = transformer.visit(tree)
-    new_file_code = compile(tree, filename, 'exec', dont_inherit=True)
+def instrument(
+    logfire: Logfire,
+    tags: Sequence[str],
+    msg_template: LiteralString | None,
+    span_name: str | None,
+    extract_args: bool,
+    allow_generator: bool,
+) -> Callable[[Callable[P, R]], Callable[P, R]]:
+    def decorator(func: Callable[P, R]) -> Callable[P, R]:
+        if getattr(func, '__code__', None) in (CONTEXTMANAGER_HELPER_CODE, ASYNCCONTEXTMANAGER_HELPER_CODE):
+            warnings.warn(
+                '@logfire.instrument should be underneath @contextlib.[async]contextmanager so that it is applied first.',
+                stacklevel=2,
+            )
 
-    # Recursively walk through the compiled code (starting from the module)
-    # to find the compiled code for the function we're instrumenting.
+        attributes = get_attributes(func, msg_template, tags)
+        final_span_name: str = span_name or attributes[ATTRIBUTES_MESSAGE_TEMPLATE_KEY]  # type: ignore
 
-    def find_code(root_code: CodeType) -> Iterator[CodeType]:
-        for const in root_code.co_consts:
-            if not isinstance(const, CodeType):
-                continue
-            matches = const.co_firstlineno == func_code.co_firstlineno and const.co_name == func_code.co_name
-            if matches:
-                yield const
-            yield from find_code(const)
+        if extract_args:
+            sig = inspect.signature(func)
 
-    [new_func_code] = find_code(new_file_code)
-    return new_func_code, logfire_name
+            def open_span(*func_args: P.args, **func_kwargs: P.kwargs):  # type: ignore
+                return logfire._instrument_span_with_args(  # type: ignore
+                    final_span_name, attributes, sig.bind(*func_args, **func_kwargs).arguments
+                )
+        else:
 
+            def open_span(*_: P.args, **__: P.kwargs):
+                return logfire._fast_span(final_span_name, attributes)  # type: ignore
+
+        if inspect.isgeneratorfunction(func):
+            if not allow_generator:
+                warnings.warn(GENERATOR_WARNING_MESSAGE, stacklevel=2)
+
+            def wrapper(*func_args: P.args, **func_kwargs: P.kwargs):  # type: ignore
+                with open_span(*func_args, **func_kwargs):
+                    yield from func(*func_args, **func_kwargs)
```

Comment:
Is this more correct? Because it avoids keeping the span on the stack when the generator is suspended?

```suggestion
                while True:
                    with open_span(*func_args, **func_kwargs):
                        value = func(*func_args, **func_kwargs)
                    yield
```

... I think there are probably things to fix like sending values and `StopIteration`.

Reply:
No, not just because it needs fixing for things, but because we generally actually want the span to be open while the generator is suspended.